File: backend/app/schemas/telemetry.py
"""Typed API schemas for observability telemetry."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class TelemetrySourceOut(BaseModel):
    id: str
    name: str
    source_type: str
    endpoint_url: str | None
    cluster_id: str | None
    is_active: bool
    config: dict | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MetricOut(BaseModel):
    id: str
    timestamp: datetime
    metric_name: str
    value: float
    unit: str
    resource_type: str
    resource_name: str | None
    source_id: str | None
    cluster_id: str | None
    namespace_name: str | None
    deployment_name: str | None
    pod_name: str | None
    service_name: str | None
    incident_id: str | None
    labels: dict | None

    model_config = {"from_attributes": True}


class MetricSeriesOut(BaseModel):
    metric_name: str
    unit: str
    points: list[MetricOut]


class LogEntryOut(BaseModel):
    id: str
    timestamp: datetime
    severity: str
    source: str
    message: str
    source_id: str | None
    cluster_id: str | None
    namespace_name: str | None
    deployment_name: str | None
    pod_name: str | None
    service_name: str | None
    incident_id: str | None
    trace_id: str | None
    span_id: str | None
    attributes: dict | None

    model_config = {"from_attributes": True}


class InfrastructureEventOut(BaseModel):
    id: str
    timestamp: datetime
    event_type: str
    reason: str
    severity: str
    message: str
    resource_type: str
    resource_name: str
    source_id: str | None
    cluster_id: str | None
    namespace_name: str | None
    deployment_name: str | None
    pod_name: str | None
    service_name: str | None
    incident_id: str | None
    attributes: dict | None

    model_config = {"from_attributes": True}


class TraceOut(BaseModel):
    id: str
    trace_id: str
    span_id: str
    parent_span_id: str | None
    operation_name: str
    service_name: str
    status: str
    start_time: datetime
    end_time: datetime
    duration_ms: int
    source_id: str | None
    cluster_id: str | None
    namespace_name: str | None
    deployment_name: str | None
    pod_name: str | None
    incident_id: str | None
    attributes: dict | None

    model_config = {"from_attributes": True}


class TelemetrySummaryOut(BaseModel):
    metrics: int
    logs: int
    events: int
    traces: int
    sources: int
    latest_timestamp: datetime | None = None


class DemoTelemetryResponse(BaseModel):
    status: str = "generated"
    clusters: int
    metrics: int
    logs: int
    events: int
    traces: int
    source: TelemetrySourceOut


class TelemetryQuery(BaseModel):
    cluster_id: str | None = None
    namespace_name: str | None = None
    deployment_name: str | None = None
    pod_name: str | None = None
    service_name: str | None = None
    incident_id: str | None = None
    limit: int = Field(default=200, ge=1, le=1000)
    metric_name: str | None = None
    severity: str | None = None
    resource_type: str | None = None
    labels: dict[str, Any] | None = None

File: frontend/src/services/telemetry.ts
import { api } from './api'

export interface TelemetrySource {
  id: string
  name: string
  source_type: string
  endpoint_url?: string | null
  cluster_id?: string | null
  is_active: boolean
  config?: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface Metric {
  id: string
  timestamp: string
  metric_name: string
  value: number
  unit: string
  resource_type: string
  resource_name?: string | null
  source_id?: string | null
  cluster_id?: string | null
  namespace_name?: string | null
  deployment_name?: string | null
  pod_name?: string | null
  service_name?: string | null
  incident_id?: string | null
  labels?: Record<string, unknown>
}

export interface LogEntry {
  id: string
  timestamp: string
  severity: string
  source: string
  message: string
  source_id?: string | null
  cluster_id?: string | null
  namespace_name?: string | null
  deployment_name?: string | null
  pod_name?: string | null
  service_name?: string | null
  incident_id?: string | null
  trace_id?: string | null
  span_id?: string | null
  attributes?: Record<string, unknown>
}

export interface InfrastructureEvent {
  id: string
  timestamp: string
  event_type: string
  reason: string
  severity: string
  message: string
  resource_type: string
  resource_name: string
  source_id?: string | null
  cluster_id?: string | null
  namespace_name?: string | null
  deployment_name?: string | null
  pod_name?: string | null
  service_name?: string | null
  incident_id?: string | null
  attributes?: Record<string, unknown>
}

export interface TraceSpan {
  id: string
  trace_id: string
  span_id: string
  parent_span_id?: string | null
  operation_name: string
  service_name: string
  status: string
  start_time: string
  end_time: string
  duration_ms: number
  source_id?: string | null
  cluster_id?: string | null
  namespace_name?: string | null
  deployment_name?: string | null
  pod_name?: string | null
  incident_id?: string | null
  attributes?: Record<string, unknown>
}

export interface TelemetrySummary {
  metrics: number
  logs: number
  events: number
  traces: number
  sources: number
  latest_timestamp?: string | null
}

export interface DemoTelemetryResponse {
  status: string
  clusters: number
  metrics: number
  logs: number
  events: number
  traces: number
  source: TelemetrySource
}

interface TelemetryParams {
  cluster_id?: string
  metric_name?: string
  severity?: string
  resource_type?: string
  namespace_name?: string
  deployment_name?: string
  pod_name?: string
  service_name?: string
  incident_id?: string
  limit?: number
}

export const telemetryApi = {
  sources: () =>
    api.get<TelemetrySource[]>('/telemetry/sources').then((r) => r.data),

  summary: () =>
    api.get<TelemetrySummary>('/telemetry/summary').then((r) => r.data),

  metrics: (params?: TelemetryParams) =>
    api.get<Metric[]>('/metrics', { params }).then((r) => r.data),

  logs: (params?: TelemetryParams) =>
    api.get<LogEntry[]>('/logs', { params }).then((r) => r.data),

  events: (params?: TelemetryParams) =>
    api.get<InfrastructureEvent[]>('/events', { params }).then((r) => r.data),

  traces: (params?: TelemetryParams) =>
    api.get<TraceSpan[]>('/traces', { params }).then((r) => r.data),

  clusterMetrics: (clusterId: string, params?: TelemetryParams) =>
    api.get<Metric[]>(`/clusters/${clusterId}/metrics`, { params }).then((r) => r.data),

  clusterLogs: (clusterId: string, params?: TelemetryParams) =>
    api.get<LogEntry[]>(`/clusters/${clusterId}/logs`, { params }).then((r) => r.data),

  clusterEvents: (clusterId: string, params?: TelemetryParams) =>
    api.get<InfrastructureEvent[]>(`/clusters/${clusterId}/events`, { params }).then((r) => r.data),

  clusterTraces: (clusterId: string, params?: TelemetryParams) =>
    api.get<TraceSpan[]>(`/clusters/${clusterId}/traces`, { params }).then((r) => r.data),

  generateDemo: () =>
    api.post<DemoTelemetryResponse>('/demo/telemetry/generate').then((r) => r.data),
}
